src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.*;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.ResourceAccessException;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+import static it.pagopa.pm.gateway.constant.ApiPaths.*;
+import static it.pagopa.pm.gateway.constant.Headers.MDC_FIELDS;
+import static it.pagopa.pm.gateway.constant.Headers.X_CLIENT_ID;
+import static it.pagopa.pm.gateway.constant.Messages.*;
+import static it.pagopa.pm.gateway.dto.enums.PaymentRequestStatusEnum.CREATED;
+import static it.pagopa.pm.gateway.dto.enums.PaymentRequestStatusEnum.DENIED;
+import static it.pagopa.pm.gateway.utils.MdcUtils.setMdcFields;
+import static java.lang.String.format;
+
+@RestController
+@Slf4j
+@RequestMapping(REQUEST_PAYMENTS_XPAY)
+public class XPayPaymentController {
+
+    private static final String APP_ORIGIN = "APP";
+    private static final String WEB_ORIGIN = "WEB";
+    private static final List<String> VALID_CLIENT_ID = Arrays.asList(APP_ORIGIN, WEB_ORIGIN);
+    private static final String EUR_CURRENCY = "978";
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    @Value("${xpay.response.urlredirect}")
+    private String PGS_RESPONSE_URL_REDIRECT;
+
+    @Value("${xpay.request.responseUrl}")
+    private String XPAY_RESPONSE_URL;
+
+    @Value("${xpay.apiKey}")
+    private String API_KEY;
+
+    @Value("${xpay.secretKey}")
+    private String SECRET_KEY;
+
+    @Autowired
+    private PaymentRequestRepository paymentRequestRepository;
+
+    @Autowired
+    private XpayService service;
+
+    @PostMapping()
+    public ResponseEntity<XPayAuthResponse> requestPaymentsXPay(@RequestHeader(value = X_CLIENT_ID) String clientId,
+                                                                @RequestHeader(required = false, value = MDC_FIELDS) String mdcFields,
+                                                                @RequestBody XPayAuthRequest pgsRequest) {
+
+        log.info("START POST - " + REQUEST_PAYMENTS_XPAY);
+        setMdcFields(mdcFields);
+        String idTransaction = pgsRequest.getIdTransaction();
+
+        if (ObjectUtils.anyNull(pgsRequest) || pgsRequest.getGrandTotal().equals(BigInteger.ZERO)) {
+            log.error("Bad Request - " + BAD_REQUEST_MSG);
+            return createxPayAuthResponse(BAD_REQUEST_MSG, HttpStatus.BAD_REQUEST, null);
+        }
+
+        if (!VALID_CLIENT_ID.contains(clientId)) {
```

Comment:
This check on the header parameter is better to be done before validating the request body